File: src/parser/immediate.rs
use miette::Result;

use crate::vm::register::Register;

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum DataSize {
    Byte,
    Word,
    DWord,
    QWord,
    Float,
    Double,
}

impl DataSize {
    pub fn size_in_bytes(&self) -> usize {
        match self {
            DataSize::Byte => 1,
            DataSize::Word => 2,
            DataSize::DWord => 4,
            DataSize::QWord => 8,
            DataSize::Float => 4,
            DataSize::Double => 8,
        }
    }
}

impl From<Register> for DataSize {
    fn from(value: Register) -> Self {
        match value {
            Register::B0 => Self::Byte,
            Register::W0 => Self::Word,
            Register::D0 => Self::DWord,
            Register::Q0 => Self::QWord,
            Register::FF0 => Self::Float,
            Register::DD0 => Self::Double,
            Register::B1 => Self::Byte,
            Register::W1 => Self::Word,
            Register::D1 => Self::DWord,
            Register::Q1 => Self::QWord,
            Register::FF1 => Self::Float,
            Register::DD1 => Self::Double,
            Register::B2 => Self::Byte,
            Register::W2 => Self::Word,
            Register::D2 => Self::DWord,
            Register::Q2 => Self::QWord,
            Register::FF2 => Self::Float,
            Register::DD2 => Self::Double,
            Register::IP => Self::QWord,
            Register::SP => Self::QWord,
            Register::BP => Self::QWord,
        }
    }
}

impl TryFrom<&str> for DataSize {
    type Error = ();

    fn try_from(value: &str) -> std::result::Result<Self, Self::Error> {
        match value.to_lowercase().as_str() {
            "byte" => Ok(Self::Byte),
            "word" => Ok(Self::Word),
            "dword" => Ok(Self::DWord),
            "qword" => Ok(Self::QWord),
            "float" => Ok(Self::Float),
            "double" => Ok(Self::Double),
            _ => Err(()),
        }
    }
}

impl TryFrom<u8> for DataSize {
    type Error = ();

    fn try_from(value: u8) -> std::result::Result<Self, Self::Error> {
        match value {
            0x00 => Ok(Self::Byte),
            0x01 => Ok(Self::Word),
            0x02 => Ok(Self::DWord),
            0x03 => Ok(Self::QWord),
            0x04 => Ok(Self::Float),
            0x05 => Ok(Self::Double),
            _ => Err(()),
        }
    }
}

impl Into<u8> for DataSize {
    fn into(self) -> u8 {
        self as u8
    }
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum Immediate {
    Byte(u8),
    Word(u16),
    DWord(u32),
    QWord(u64),
    Float(f32),
    Double(f64),
}

impl Immediate {
    pub fn as_u8(self) -> Result<u8> {
        match self {
            Immediate::Byte(v) => Ok(v),
            Immediate::Word(v) => Ok(v as u8),
            Immediate::DWord(v) => Ok(v as u8),
            Immediate::QWord(v) => Ok(v as u8),
            Immediate::Float(v) => Ok(v as u8),
            Immediate::Double(v) => Ok(v as u8),
        }
    }

    pub fn as_u16(self) -> Result<u16> {
        match self {
            Immediate::Byte(v) => Ok(v as u16),
            Immediate::Word(v) => Ok(v),
            Immediate::DWord(v) => Ok(v as u16),
            Immediate::QWord(v) => Ok(v as u16),
            Immediate::Float(v) => Ok(v as u16),
            Immediate::Double(v) => Ok(v as u16),
        }
    }

    pub fn as_u32(self) -> Result<u32> {
        match self {
            Immediate::Byte(v) => Ok(v as u32),
            Immediate::Word(v) => Ok(v as u32),
            Immediate::DWord(v) => Ok(v),
            Immediate::QWord(v) => Ok(v as u32),
            Immediate::Float(v) => Ok(v as u32),
            Immediate::Double(v) => Ok(v as u32),
        }
    }

    pub fn as_u64(self) -> Result<u64> {
        match self {
            Immediate::Byte(v) => Ok(v as u64),
            Immediate::Word(v) => Ok(v as u64),
            Immediate::DWord(v) => Ok(v as u64),
            Immediate::QWord(v) => Ok(v),
            Immediate::Float(v) => Ok(v as u64),
            Immediate::Double(v) => Ok(v as u64),
        }
    }

    pub fn as_f32(self) -> Result<f32> {
        match self {
            Immediate::Byte(v) => Ok(v as f32),
            Immediate::Word(v) => Ok(v as f32),
            Immediate::DWord(v) => Ok(v as f32),
            Immediate::QWord(v) => Ok(v as f32),
            Immediate::Float(v) => Ok(v),
            Immediate::Double(v) => Ok(v as f32),
        }
    }

    pub fn as_f64(self) -> Result<f64> {
        match self {
            Immediate::Byte(v) => Ok(v as f64),
            Immediate::Word(v) => Ok(v as f64),
            Immediate::DWord(v) => Ok(v as f64),
            Immediate::QWord(v) => Ok(v as f64),
            Immediate::Float(v) => Ok(v as f64),
            Immediate::Double(v) => Ok(v),
        }
    }

    pub fn as_usize(self) -> Result<usize> {
        match self {
            Immediate::Byte(v) => Ok(v as usize),
            Immediate::Word(v) => Ok(v as usize),
            Immediate::DWord(v) => Ok(v as usize),
            Immediate::QWord(v) => Ok(v as usize),
            Immediate::Float(v) => Ok(v as usize),
            Immediate::Double(v) => Ok(v as usize),
        }
    }

    pub fn size(&self) -> DataSize {
        match self {
            Immediate::Byte(_) => DataSize::Byte,
            Immediate::Word(_) => DataSize::Word,
            Immediate::DWord(_) => DataSize::DWord,
            Immediate::QWord(_) => DataSize::QWord,
            Immediate::Float(_) => DataSize::Float,
            Immediate::Double(_) => DataSize::Double,
        }
    }
}

File: src/parser/immediate.zig
const std = @import("std");
const mem = std.mem;
const ascii = std.ascii;
const Register = @import("../vm/register.zig").Register;

pub const DataSize = enum {
    byte,
    word,
    dword,
    qword,
    float,
    double,

    pub fn sizeInBytes(self: DataSize) usize {
        return switch (self) {
            .byte => 1,
            .word => 2,
            .dword => 4,
            .qword => 8,
            .float => 4,
            .double => 8,
        };
    }

    pub fn fromRegister(value: Register) DataSize {
        return switch (value) {
            .b0, .b1, .b2, .b3, .b4, .b5, .b6, .b7, .b8, .b9, .b10, .b11, .b12, .b13, .b14, .b15 => .byte,
            .w0, .w1, .w2, .w3, .w4, .w5, .w6, .w7, .w8, .w9, .w10, .w11, .w12, .w13, .w14, .w15 => .word,
            .d0, .d1, .d2, .d3, .d4, .d5, .d6, .d7, .d8, .d9, .d10, .d11, .d12, .d13, .d14, .d15 => .dword,
            .q0, .q1, .q2, .q3, .q4, .q5, .q6, .q7, .q8, .q9, .q10, .q11, .q12, .q13, .q14, .q15, .ip, .sp, .bp => .qword,
            .ff0, .ff1, .ff2, .ff3, .ff4, .ff5, .ff6, .ff7, .ff8, .ff9, .ff10, .ff11, .ff12, .ff13, .ff14, .ff15 => .float,
            .dd0, .dd1, .dd2, .dd3, .dd4, .dd5, .dd6, .dd7, .dd8, .dd9, .dd10, .dd11, .dd12, .dd13, .dd14, .dd15 => .double,
        };
    }

    pub fn fromString(value: []const u8) !DataSize {
        if (ascii.eqlIgnoreCase(value, "byte")) return .byte;
        if (ascii.eqlIgnoreCase(value, "word")) return .word;
        if (ascii.eqlIgnoreCase(value, "dword")) return .dword;
        if (ascii.eqlIgnoreCase(value, "qword")) return .qword;
        if (ascii.eqlIgnoreCase(value, "float")) return .float;
        if (ascii.eqlIgnoreCase(value, "double")) return .double;
        return error.InvalidDataSize;
    }

    pub fn fromU8(value: u8) !DataSize {
        return switch (value) {
            0x00 => .byte,
            0x01 => .word,
            0x02 => .dword,
            0x03 => .qword,
            0x04 => .float,
            0x05 => .double,
            else => error.UnknownDataSize,
        };
    }

    pub fn toU8(self: DataSize) u8 {
        return @intFromEnum(self);
    }
};

pub const Immediate = union(enum) {
    byte: u8,
    word: u16,
    dword: u32,
    qword: u64,
    float: f32,
    double: f64,

    pub fn asU8(self: Immediate) u8 {
        return switch (self) {
            .byte => |v| v,
            .word => |v| @truncate(v),
            .dword => |v| @truncate(v),
            .qword => |v| @truncate(v),
            .float => |v| @truncate(@as(u8, @intFromFloat(v))),
            .double => |v| @truncate(@as(u8, @intFromFloat(v))),
        };
    }

    pub fn asU16(self: Immediate) u16 {
        return switch (self) {
            .byte => |v| @intCast(v),
            .word => |v| v,
            .dword => |v| @truncate(v),
            .qword => |v| @truncate(v),
            .float => |v| @truncate(@as(u16, @intFromFloat(v))),
            .double => |v| @truncate(@as(u16, @intFromFloat(v))),
        };
    }

    pub fn asU32(self: Immediate) u32 {
        return switch (self) {
            .byte => |v| @intCast(v),
            .word => |v| @intCast(v),
            .dword => |v| v,
            .qword => |v| @truncate(v),
            .float => |v| @truncate(@as(u32, @intFromFloat(v))),
            .double => |v| @truncate(@as(u32, @intFromFloat(v))),
        };
    }

    pub fn asU64(self: Immediate) u64 {
        return switch (self) {
            .byte => |v| @intCast(v),
            .word => |v| @intCast(v),
            .dword => |v| @intCast(v),
            .qword => |v| v,
            .float => |v| @truncate(@as(u64, @intFromFloat(v))),
            .double => |v| @truncate(@as(u64, @intFromFloat(v))),
        };
    }

    pub fn asF32(self: Immediate) f32 {
        return switch (self) {
            .byte => |v| @floatFromInt(v),
            .word => |v| @floatFromInt(v),
            .dword => |v| @floatFromInt(v),
            .qword => |v| @floatFromInt(v),
            .float => |v| v,
            .double => |v| @floatCast(v),
        };
    }

    pub fn asF64(self: Immediate) f64 {
        return switch (self) {
            .byte => |v| @floatFromInt(v),
            .word => |v| @floatFromInt(v),
            .dword => |v| @floatFromInt(v),
            .qword => |v| @floatFromInt(v),
            .float => |v| @floatCast(v),
            .double => |v| v,
        };
    }

    pub fn asUsize(self: Immediate) usize {
        return switch (self) {
            .byte => |v| @intCast(v),
            .word => |v| @intCast(v),
            .dword => |v| @intCast(v),
            .qword => |v| @intCast(v),
            .float => |v| @intFromFloat(v),
            .double => |v| @intFromFloat(v),
        };
    }

    pub fn size(self: Immediate) DataSize {
        return switch (self) {
            .byte => .byte,
            .word => .word,
            .dword => .dword,
            .qword => .qword,
            .float => .float,
            .double => .double,
        };
    }

    pub fn eql(self: Immediate, other: Immediate) bool {
        return switch (self) {
            .byte => |v| other == .byte and v == other.byte,
            .word => |v| other == .word and v == other.word,
            .dword => |v| other == .dword and v == other.dword,
            .qword => |v| other == .qword and v == other.qword,
            .float => |v| other == .float and v == other.float,
            .double => |v| other == .double and v == other.double,
        };
    }

    pub fn lessThan(self: Immediate, other: Immediate) bool {
        return switch (self) {
            .byte => |v| other == .byte and v < other.byte,
            .word => |v| other == .word and v < other.word,
            .dword => |v| other == .dword and v < other.dword,
            .qword => |v| other == .qword and v < other.qword,
            .float => |v| other == .float and v < other.float,
            .double => |v| other == .double and v < other.double,
        };
    }
};
